File: plugin/Source/Pipeline.cpp
#include "Pipeline.h"
#include "AbletonClient.h"
#include <juce_audio_formats/juce_audio_formats.h>
#include <juce_events/juce_events.h>

namespace voicectrl
{
namespace
{
constexpr const char* kWhisperCli = "/opt/homebrew/bin/whisper-cli";
constexpr const char* kModelPath  = VOICECTRL_RESOURCES_DIR "/models/ggml-base.en.bin";
constexpr const char* kOllamaUrl  = "http://127.0.0.1:11434/api/chat";
constexpr const char* kModel      = "qwen2.5:7b-instruct";

const char* kSystem =
    "You are VoiceCtrl, a real-time voice assistant for Ableton Live.\n"
    "Translate the user's spoken instruction into one or more tool calls.\n"
    "Rules: track and scene indices start at 0 (track 1 == index 0). "
    "Volumes are dB (0 = unity). Be decisive; never ask follow-up questions.";

// Build the tool list once.
juce::var buildTools()
{
    auto fn = [] (juce::String name, juce::String desc,
                  std::vector<std::pair<juce::String, std::pair<juce::String, juce::String>>> props = {},
                  std::vector<juce::String> required = {}) -> juce::var
    {
        auto* params = new juce::DynamicObject();
        params->setProperty ("type", "object");
        auto* propsObj = new juce::DynamicObject();
        for (auto& [k, v] : props)
        {
            auto* p = new juce::DynamicObject();
            p->setProperty ("type", v.first);
            p->setProperty ("description", v.second);
            propsObj->setProperty (k, juce::var (p));
        }
        params->setProperty ("properties", juce::var (propsObj));
        if (! required.empty())
        {
            juce::Array<juce::var> req;
            for (auto& r : required) req.add (r);
            params->setProperty ("required", juce::var (req));
        }
        auto* function = new juce::DynamicObject();
        function->setProperty ("name", name);
        function->setProperty ("description", desc);
        function->setProperty ("parameters", juce::var (params));
        auto* tool = new juce::DynamicObject();
        tool->setProperty ("type", "function");
        tool->setProperty ("function", juce::var (function));
        return juce::var (tool);
    };

    juce::Array<juce::var> tools;
    tools.add (fn ("start_playback", "Start playback in Ableton Live."));
    tools.add (fn ("stop_playback",  "Stop playback in Ableton Live."));
    tools.add (fn ("set_tempo", "Set the song tempo in BPM.",
                   {{ "tempo", { "number", "BPM 20-999" }}}, { "tempo" }));
    tools.add (fn ("set_metronome", "Enable or disable the metronome.",
                   {{ "enabled", { "boolean", "" }}}, { "enabled" }));
    tools.add (fn ("start_recording", "Start arrangement recording."));
    tools.add (fn ("stop_recording",  "Stop recording."));
    tools.add (fn ("toggle_session_record", "Toggle the session record button."));
    tools.add (fn ("capture_midi", "Capture recently played MIDI into a clip."));
    tools.add (fn ("create_midi_track", "Create a new MIDI track at index (-1 = end).",
                   {{ "index", { "integer", "" }}}));
    tools.add (fn ("create_audio_track", "Create a new audio track at index (-1 = end).",
                   {{ "index", { "integer", "" }}}));
    tools.add (fn ("set_track_volume", "Set track volume in dB.",
                   {{ "track_index", { "integer", "" }}, { "volume_db", { "number", "dB" }}},
                   { "track_index", "volume_db" }));
    tools.add (fn ("set_track_pan", "Set track pan (-1..+1).",
                   {{ "track_index", { "integer", "" }}, { "pan", { "number", "-1..+1" }}},
                   { "track_index", "pan" }));
    tools.add (fn ("set_track_mute", "Mute or unmute a track.",
                   {{ "track_index", { "integer", "" }}, { "muted", { "boolean", "" }}},
                   { "track_index", "muted" }));
    tools.add (fn ("set_track_solo", "Solo or unsolo a track.",
                   {{ "track_index", { "integer", "" }}, { "soloed", { "boolean", "" }}},
                   { "track_index", "soloed" }));
    tools.add (fn ("set_track_arm", "Arm or disarm a track.",
                   {{ "track_index", { "integer", "" }}, { "armed", { "boolean", "" }}},
                   { "track_index", "armed" }));
    tools.add (fn ("select_track", "Select a track by index.",
                   {{ "track_index", { "integer", "" }}}, { "track_index" }));
    tools.add (fn ("fire_clip", "Fire a session-view clip.",
                   {{ "track_index", { "integer", "" }}, { "clip_index", { "integer", "" }}},
                   { "track_index", "clip_index" }));
    tools.add (fn ("stop_clip", "Stop a session-view clip.",
                   {{ "track_index", { "integer", "" }}, { "clip_index", { "integer", "" }}},
                   { "track_index", "clip_index" }));
    tools.add (fn ("fire_scene", "Fire a scene.",
                   {{ "scene_index", { "integer", "" }}}, { "scene_index" }));
    tools.add (fn ("set_arrangement_loop", "Set arrangement loop region in beats.",
                   {{ "start", { "number", "" }}, { "end", { "number", "" }}, { "enabled", { "boolean", "" }}},
                   { "start", "end" }));
    tools.add (fn ("jump_to_time", "Jump the playhead to a time in beats.",
                   {{ "time", { "number", "" }}}, { "time" }));
    tools.add (fn ("get_session_info", "Read the current Live set."));
    return juce::var (tools);
}

double dbToLive (double db)
{
    if (db <= -70.0) return 0.0;
    return juce::jlimit (0.0, 1.0, std::pow (10.0, (db - 6.0) / 30.0) * std::pow (10.0, 6.0 / 30.0));
}

// Translate an LLM tool call into the AbletonMCP (type, params) pair.
std::pair<juce::String, juce::var> translate (const juce::String& name, const juce::var& a)
{
    auto* o = new juce::DynamicObject();
    auto get   = [&] (const char* k) -> juce::var { return a[k]; };
    auto setI  = [&] (const char* k, int v)        { o->setProperty (k, v); };
    auto setD  = [&] (const char* k, double v)     { o->setProperty (k, v); };
    auto setB  = [&] (const char* k, bool v)       { o->setProperty (k, v); };

    if (name == "start_playback")        return { "start_playback",        juce::var (o) };
    if (name == "stop_playback")         return { "stop_playback",         juce::var (o) };
    if (name == "start_recording")       return { "start_recording",       juce::var (o) };
    if (name == "stop_recording")        return { "stop_recording",        juce::var (o) };
    if (name == "toggle_session_record") return { "toggle_session_record", juce::var (o) };
    if (name == "capture_midi")          return { "capture_midi",          juce::var (o) };
    if (name == "get_session_info")      return { "get_session_info",      juce::var (o) };

    if (name == "set_tempo")        { setD ("tempo",   (double) get ("tempo")); return { name, juce::var (o) }; }
    if (name == "set_metronome")    { setB ("enabled", (bool)   get ("enabled")); return { name, juce::var (o) }; }
    if (name == "create_midi_track" || name == "create_audio_track")
    { setI ("index", get ("index").isVoid() ? -1 : (int) get ("index")); return { name, juce::var (o) }; }

    if (name == "set_track_volume")
    {
        setI ("track_index", (int) get ("track_index"));
        setD ("volume", dbToLive ((double) get ("volume_db")));
        return { name, juce::var (o) };
    }
    if (name == "set_track_pan")  { setI ("track_index", (int) get ("track_index")); setD ("pan",  (double) get ("pan"));  return { name, juce::var (o) }; }
    if (name == "set_track_mute") { setI ("track_index", (int) get ("track_index")); setB ("mute", (bool)   get ("muted")); return { name, juce::var (o) }; }
    if (name == "set_track_solo") { setI ("track_index", (int) get ("track_index")); setB ("solo", (bool)   get ("soloed")); return { name, juce::var (o) }; }
    if (name == "set_track_arm")  { setI ("track_index", (int) get ("track_index")); setB ("arm",  (bool)   get ("armed"));  return { name, juce::var (o) }; }
    if (name == "select_track")   { setI ("track_index", (int) get ("track_index")); return { name, juce::var (o) }; }
    if (name == "fire_clip" || name == "stop_clip")
    { setI ("track_index", (int) get ("track_index")); setI ("clip_index", (int) get ("clip_index")); return { name, juce::var (o) }; }
    if (name == "fire_scene")     { setI ("scene_index", (int) get ("scene_index")); return { name, juce::var (o) }; }
    if (name == "jump_to_time")   { setD ("time", (double) get ("time")); return { name, juce::var (o) }; }
    if (name == "set_arrangement_loop")
    {
        setD ("start", (double) get ("start"));
        setD ("end",   (double) get ("end"));
        setB ("enabled", get ("enabled").isVoid() ? true : (bool) get ("enabled"));
        return { name, juce::var (o) };
    }

    delete o;
    return { {}, juce::var() };
}
}

// ──────────────────────────────────────────────────────────────────────────
Pipeline::Pipeline() : juce::Thread ("VoiceCtrl-Pipeline")
{
    startThread();
}

Pipeline::~Pipeline()
{
    stopThread (2000);
    hasWork.signal();
}

void Pipeline::submit (juce::AudioBuffer<float> mono16k)
{
    {
        const juce::ScopedLock sl (lock);
        queue.emplace_back (std::move (mono16k));
    }
    hasWork.signal();
}

void Pipeline::run()
{
    while (! threadShouldExit())
    {
        pollHealth();

        juce::AudioBuffer<float> buf;
        bool got = false;
        {
            const juce::ScopedLock sl (lock);
            if (! queue.empty()) { buf = std::move (queue.front()); queue.pop_front(); got = true; }
        }
        if (! got) { hasWork.wait (500); continue; }
        try { process (buf); }
        catch (std::exception& e) { post (juce::String ("error: ") + e.what()); }
    }
}

void Pipeline::pollHealth()
{
    const auto now = juce::Time::getMillisecondCounter();
    if (now - lastHealthMs < 5000) return;
    lastHealthMs = now;
    abletonUp.store (AbletonClient::healthCheck());

    juce::URL u ("http://127.0.0.1:11434/api/tags");
    if (auto stream = u.createInputStream (juce::URL::InputStreamOptions (juce::URL::ParameterHandling::inAddress)
                                              .withConnectionTimeoutMs (1000)))
        ollamaUp.store (true);
    else
        ollamaUp.store (false);
}

void Pipeline::post (const juce::String& s)
{
    if (logger)
        juce::MessageManager::callAsync ([fn = logger, s] { fn (s); });
}

void Pipeline::process (juce::AudioBuffer<float>& buf)
{
    auto tmpDir = juce::File::getSpecialLocation (juce::File::tempDirectory)
                      .getChildFile ("voicectrl-" + juce::String::toHexString ((juce::int64) juce::Time::currentTimeMillis()));
    tmpDir.createDirectory();

    auto wavFile = tmpDir.getChildFile ("in.wav");
    {
        juce::WavAudioFormat fmt;
        std::unique_ptr<juce::FileOutputStream> os (wavFile.createOutputStream());
        if (os)
        {
            std::unique_ptr<juce::AudioFormatWriter> w (
                fmt.createWriterFor (os.get(), 16000.0, 1, 16, {}, 0));
            if (w)
            {
                os.release();
                w->writeFromAudioSampleBuffer (buf, 0, buf.getNumSamples());
            }
        }
    }
    if (! wavFile.existsAsFile() || wavFile.getSize() < 100)
    { post ("STT: failed to write wav"); tmpDir.deleteRecursively(); return; }

    post ("transcribing…");
    auto transcript = transcribe (wavFile);
    tmpDir.deleteRecursively();
    if (transcript.isEmpty()) { post ("(no speech detected)"); return; }
    post ("you: " + transcript);

    post ("planning…");
    auto reply = askOllama (transcript);
    if (reply.isVoid()) { post ("LLM error"); return; }

    executeToolCalls (reply, transcript);
}

juce::String Pipeline::transcribe (const juce::File& wav)
{
    juce::ChildProcess proc;
    juce::StringArray args;
    args.add (kWhisperCli);
    args.add ("-m"); args.add (kModelPath);
    args.add ("-f"); args.add (wav.getFullPathName());
    args.add ("-nt");
    args.add ("-otxt");
    args.add ("-of"); args.add (wav.withFileExtension ("").getFullPathName());
    if (! proc.start (args)) { post ("whisper-cli failed to start"); return {}; }
    proc.waitForProcessToFinish (60000);
    auto txt = wav.withFileExtension ("txt");
    if (! txt.existsAsFile()) return proc.readAllProcessOutput().trim();
    return txt.loadFileAsString().trim();
}

juce::var Pipeline::askOllama (const juce::String& transcript)
{
    static const juce::var TOOLS = buildTools();

    auto* sys = new juce::DynamicObject();
    sys->setProperty ("role", "system"); sys->setProperty ("content", juce::String (kSystem));
    auto* usr = new juce::DynamicObject();
    usr->setProperty ("role", "user");   usr->setProperty ("content", transcript);

    juce::Array<juce::var> messages; messages.add (juce::var (sys)); messages.add (juce::var (usr));

    auto* opts = new juce::DynamicObject();
    opts->setProperty ("temperature", 0.1);

    auto* body = new juce::DynamicObject();
    body->setProperty ("model",    juce::String (kModel));
    body->setProperty ("stream",   false);
    body->setProperty ("messages", juce::var (messages));
    body->setProperty ("tools",    TOOLS);
    body->setProperty ("options",  juce::var (opts));

    juce::URL url (kOllamaUrl);
    url = url.withPOSTData (juce::JSON::toString (juce::var (body)));

    auto opt = juce::URL::InputStreamOptions (juce::URL::ParameterHandling::inAddress)
                   .withConnectionTimeoutMs (120000)
                   .withExtraHeaders ("Content-Type: application/json")
                   .withHttpRequestCmd ("POST");
    auto in = url.createInputStream (opt);
    if (! in) { post ("Ollama: connect failed"); return {}; }
    auto txt = in->readEntireStreamAsString();
    return juce::JSON::parse (txt);
}

void Pipeline::executeToolCalls (const juce::var& reply, const juce::String& transcript)
{
    juce::ignoreUnused (transcript);
    auto msg = reply["message"];
    auto calls = msg["tool_calls"];
    if (! calls.isArray() || calls.size() == 0)
    {
        auto txt = msg["content"].toString().trim();
        if (txt.isNotEmpty()) post (txt);
        else                  post ("(no tool calls)");
        return;
    }

    for (auto& c : *calls.getArray())
    {
        auto fn = c["function"];
        auto name = fn["name"].toString();
        juce::var args = fn["arguments"];
        if (args.isString()) args = juce::JSON::parse (args.toString());
        auto [cmdType, params] = translate (name, args);
        if (cmdType.isEmpty()) { post ("✗ unknown tool: " + name); continue; }
        auto resp = AbletonClient::send (cmdType, params);
        if (auto* ro = resp.getDynamicObject())
            post ("✓ " + name + " → " + ro->getProperty ("status").toString());
        else
            post ("✗ " + name + " (no response from Ableton)");
    }
}
}

File: src/pipeline.js
"use strict";

const config = require("./config");
const logger    = require("./logger");
const { abletonSend } = require("./services/ableton");
const { plan, extractToolCalls } = require("./services/ollama");
const { transcribe } = require("./services/whisper");

// ── validation helpers ─────────────────────────────────────────────────────
function dbToLive(db) {
  if (db <= -70) return 0;
  return Math.min(1, Math.max(0, 0.85 * Math.pow(10, db / 20)));
}

function finiteNumber(v, name, min = -Infinity, max = Infinity) {
  const n = Number(v);
  if (!Number.isFinite(n)) throw new Error(`invalid ${name}: ${JSON.stringify(v)}`);
  if (n < min || n > max) throw new Error(`${name} out of range: ${n} (${min}..${max})`);
  return n;
}

function intValue(v, name, min = -1, max = 4096) {
  const n = finiteNumber(v, name, min, max);
  if (!Number.isInteger(n)) throw new Error(`${name} must be integer: ${n}`);
  return n;
}

function boolValue(v, name) {
  if (typeof v === "boolean") return v;
  if (typeof v === "string") {
    const s = v.trim().toLowerCase();
    if (["true", "on", "yes", "1"].includes(s)) return true;
    if (["false", "off", "no", "0"].includes(s)) return false;
  }
  if (v === 1) return true;
  if (v === 0) return false;
  throw new Error(`invalid ${name}: ${JSON.stringify(v)}`);
}

// ── command map ────────────────────────────────────────────────────────────
const COMMAND_MAP = {
  start_playback:        () => ["start_playback", {}],
  stop_playback:         () => ["stop_playback", {}],
  set_tempo:             a  => ["set_tempo", { tempo: finiteNumber(a.tempo, "tempo", 20, 999) }],
  set_metronome:         a  => ["set_metronome", { enabled: boolValue(a.enabled, "enabled") }],
  start_recording:       () => ["start_recording", {}],
  stop_recording:        () => ["stop_recording", {}],
  toggle_session_record: () => ["toggle_session_record", {}],
  capture_midi:          () => ["capture_midi", {}],
  create_midi_track:     a  => ["create_midi_track", { index: a.index === undefined || a.index === null ? -1 : intValue(a.index, "index", -1) }],
  create_audio_track:    a  => ["create_audio_track", { index: a.index === undefined || a.index === null ? -1 : intValue(a.index, "index", -1) }],
  set_track_volume:      a  => ["set_track_volume", { track_index: intValue(a.track_index, "track_index", 0), volume: dbToLive(finiteNumber(a.volume_db, "volume_db", -70, 6)) }],
  set_track_pan:         a  => ["set_track_pan",    { track_index: intValue(a.track_index, "track_index", 0), pan: finiteNumber(a.pan, "pan", -1, 1) }],
  set_track_mute:        a  => ["set_track_mute",   { track_index: intValue(a.track_index, "track_index", 0), mute: boolValue(a.muted, "muted") }],
  set_track_solo:        a  => ["set_track_solo",   { track_index: intValue(a.track_index, "track_index", 0), solo: boolValue(a.soloed, "soloed") }],
  set_track_arm:         a  => ["set_track_arm",    { track_index: intValue(a.track_index, "track_index", 0), arm:  boolValue(a.armed, "armed") }],
  select_track:          a  => ["select_track",     { track_index: intValue(a.track_index, "track_index", 0) }],
  fire_clip:             a  => ["fire_clip",        { track_index: intValue(a.track_index, "track_index", 0), clip_index: intValue(a.clip_index, "clip_index", 0) }],
  stop_clip:             a  => ["stop_clip",        { track_index: intValue(a.track_index, "track_index", 0), clip_index: intValue(a.clip_index, "clip_index", 0) }],
  fire_scene:            a  => ["fire_scene",       { scene_index: intValue(a.scene_index, "scene_index", 0) }],
  set_arrangement_loop:  a  => ["set_arrangement_loop", { start: finiteNumber(a.start, "start", 0), end: finiteNumber(a.end, "end", 0), enabled: a.enabled === undefined ? true : boolValue(a.enabled, "enabled") }],
  jump_to_time:          a  => ["jump_to_time",     { time: finiteNumber(a.time, "time", 0) }],
  get_session_info:      () => ["get_session_info", {}],
};

const UPSTREAM_ABLETON_COMMANDS = new Set([
  "get_session_info", "get_track_info", "create_midi_track", "set_track_name",
  "create_clip", "add_notes_to_clip", "set_clip_name", "set_tempo",
  "fire_clip", "stop_clip", "start_playback", "stop_playback",
  "load_browser_item",
]);

// ── direct dispatch patterns ────────────────────────────────────────────────
// Simple commands matched by regex — no LLM needed.
// Returns { tool, args } or null (falls through to LLM).
function tryDirectParse(text) {
  const t = text.toLowerCase().trim();
  const num = (str) => {
    const m = str.match(/[\d.]+/);
    return m ? parseFloat(m[0]) : null;
  };
  const trackNum = (str) => {
    const m = str.match(/(?:track |tr )?(\d+)/i);
    return m ? parseInt(m[1]) - 1 : 0; // 1-based → 0-based
  };

  // Transport
  if (/\b(play|start|begin|resume|go)\b/i.test(t) && !/\b(stop|pause)\b/i.test(t))
    return { tool: "start_playback", args: {} };
  if (/\b(stop|pause|halt)\b/i.test(t) && !/\b(stop clip|stop recording)\b/i.test(t))
    return { tool: "stop_playback", args: {} };
  if (/\b(stop recording|stop record)\b/i.test(t))
    return { tool: "stop_recording", args: {} };
  if (/\brecord\b/i.test(t) && !/\bstop\b/i.test(t))
    return { tool: "start_recording", args: {} };
  if (/\b(stop recording|stop record)\b/i.test(t))
    return { tool: "stop_recording", args: {} };
  if (/\b(session record|toggle session record)\b/i.test(t))
    return { tool: "toggle_session_record", args: {} };
  if (/\bcapture midi\b/i.test(t))
    return { tool: "capture_midi", args: {} };

  // Metronome
  if (/\b(metronome|click)\b/i.test(t)) {
    const on  = /\b(on|enable|start)\b/i.test(t) && !/\boff\b/i.test(t);
    const off = /\b(off|disable|stop)\b/i.test(t);
    return { tool: "set_metronome", args: { enabled: on && !off } };
  }

  // Tempo — "set tempo to 120", "120 BPM", "tempo 120", "set the tempo to 120"
  if (/\btempo\b/i.test(t) || /\b\bpm\b/i.test(t) || /^\d{3}\b/.test(t)) {
    const n = t.match(/(?:to |the tempo |tempo )?(\d{3})/i)?.[1]
              || t.match(/(\d{3})\s*bpm/i)?.[1]
              || num(t);
    if (n !== null) return { tool: "set_tempo", args: { tempo: parseInt(n) } };
  }

  // Track mute/solo/arm — "mute track 2", "unsolo track 1", "disarm track 3"
  const muteMatch = t.match(/\b(mute|unmute)\b.*(?:track |tr )?(\d+)/i);
  if (muteMatch) return { tool: "set_track_mute", args: { track_index: parseInt(muteMatch[2]) - 1, muted: !/\bunmute\b/i.test(t) } };
  const soloMatch = t.match(/\b(solo|unsolo)\b.*(?:track |tr )?(\d+)/i);
  if (soloMatch) return { tool: "set_track_solo", args: { track_index: parseInt(soloMatch[2]) - 1, soloed: !/\bunsolo\b/i.test(t) } };
  const armMatch  = t.match(/\b(arm|disarm)\b.*(?:track |tr )?(\d+)/i);
  if (armMatch)   return { tool: "set_track_arm",   args: { track_index: parseInt(armMatch[2]) - 1, armed: !/\bdisarm\b/i.test(t) } };

  // Track volume — "track 2 volume -6", "set track 1 to -3dB"
  const volMatch = t.match(/(?:track |tr )(\d+).*volume.*?(?:to\s*)?(-?\d+(?:\.\d+)?)\s*(?:db|dB)?$/i)
                 || t.match(/volume.*?(?:track |tr )(\d+).*?(?:to\s*)?(-?\d+(?:\.\d+)?)\s*(?:db|dB)?$/i);
  if (volMatch) return { tool: "set_track_volume", args: { track_index: parseInt(volMatch[1]) - 1, volume_db: parseFloat(volMatch[2]) } };

  // Track pan — "pan track 2 left", "set track 1 pan to 0.5"
  if (/\bpan\b/i.test(t)) {
    const panLeft  = /\bleft\b/i.test(t) ? -1 : null;
    const panRight = /\bright\b/i.test(t) ? 1  : null;
    const panVal   = t.match(/pan\s+(?:to\s*)?(-?\d+(?:\.\d+)?)/i)?.[1];
    const tIdx     = t.match(/(?:track |tr )(\d+)/i)?.[1];
    if (panLeft !== null || panRight !== null)
      return { tool: "set_track_pan", args: { track_index: tIdx ? parseInt(tIdx) - 1 : 0, pan: panLeft ?? panRight } };
    if (panVal !== null)
      return { tool: "set_track_pan", args: { track_index: tIdx ? parseInt(tIdx) - 1 : 0, pan: parseFloat(panVal) } };
  }

  // Scene — "fire scene 1", "launch scene 3"
  const sceneMatch = t.match(/(?:fire|launch|play)\s+scene\s+(\d+)/i);
  if (sceneMatch) return { tool: "fire_scene", args: { scene_index: parseInt(sceneMatch[1]) - 1 } };

  // Clip — "fire clip 2 on track 3", "stop clip 1 track 2"
  const fireClipMatch = t.match(/fire.*clip\s+(\d+).*(?:track |tr )(\d+)/i)
                      || t.match(/(?:track |tr )(\d+).*fire.*clip\s+(\d+)/i);
  if (fireClipMatch) return { tool: "fire_clip", args: { track_index: parseInt(fireClipMatch[1]) - 1, clip_index: parseInt(fireClipMatch[2]) - 1 } };

  const stopClipMatch = t.match(/stop.*clip\s+(\d+).*(?:track |tr )(\d+)/i)
                     || t.match(/(?:track |tr )(\d+).*stop.*clip\s+(\d+)/i);
  if (stopClipMatch) return { tool: "stop_clip", args: { track_index: parseInt(stopClipMatch[1]) - 1, clip_index: parseInt(stopClipMatch[2]) - 1 } };

  // Create track — "create a midi track", "add audio track"
  if (/\b(create|add|make)\b/i.test(t)) {
    if (/\bmidi\b/i.test(t)) return { tool: "create_midi_track",  args: { index: -1 } };
    if (/\baudio\b/i.test(t)) return { tool: "create_audio_track", args: { index: -1 } };
  }

  // Loop — "loop 1 to 4", "set loop from 2 to 8"
  const loopMatch = t.match(/(?:loop|set loop).*?(\d+(?:\.\d+)?).*(?:to|until|end).*?(\d+(?:\.\d+)?)/i);
  if (loopMatch) return { tool: "set_arrangement_loop", args: { start: parseFloat(loopMatch[1]), end: parseFloat(loopMatch[2]), enabled: true } };

  // Jump — "jump to 16", "go to beat 32"
  if (/\b(jump|go to|seek)\b.*?(\d+(?:\.\d+)?)/i.test(t)) {
    const j = t.match(/(\d+(?:\.\d+)?)/);
    if (j) return { tool: "jump_to_time", args: { time: parseFloat(j[1]) } };
  }

  return null; // fall through to LLM
}

// ── execute a single command ───────────────────────────────────────────────
async function execOne(toolName, args, rid) {
  const t = COMMAND_MAP[toolName];
  if (!t) return { tool: toolName, args, error: "unknown tool" };
  try {
    const [type, params] = t(args);
    if (!UPSTREAM_ABLETON_COMMANDS.has(type)) {
      logger.warn(`[${rid}] command may need extended AbletonMCP: ${type}`, rid);
    }
    logger.info(`[${rid}] → ableton: ${type} ${JSON.stringify(params)}`, rid);
    const r = await abletonSend(type, params);
    logger.info(`[${rid}] ← ableton: ${type} ok`, rid);
    return { tool: toolName, args, ableton: r };
  } catch (e) {
    logger.error(`[${rid}] tool ${toolName} failed: ${e.message}`, rid);
    return { tool: toolName, args, error: e.message };
  }
}

// ── main pipeline ──────────────────────────────────────────────────────────
async function processAudio(audioBytes, rid) {
  const started = Date.now();
  rid = rid || logger.rid();
  logger.info(`[${rid}] listen: audio ${audioBytes.length} bytes`, rid);

  const transcript = await transcribe(audioBytes, rid);
  if (!transcript) {
    return { request_id: rid, transcript: "", actions: [], assistant: "(no speech detected)", duration_ms: Date.now() - started };
  }
  logger.info(`[${rid}] transcript: "${transcript}"`, rid);
  return runCommand(transcript, rid, started);
}

async function processCommand(text, rid) {
  const started = Date.now();
  rid = rid || logger.rid();
  logger.info(`[${rid}] command: "${text}"`, rid);
  return runCommand(text, rid, started);
}

async function runCommand(text, rid, started) {
  // 1. Try direct dispatch (no LLM latency)
  const direct = tryDirectParse(text);
  if (direct) {
    logger.info(`[${rid}] direct dispatch: ${direct.tool}`, rid);
    const result = await execOne(direct.tool, direct.args, rid);
    return {
      request_id: rid,
      transcript: text,
      actions: [result],
      assistant: result.error ? "" : `${direct.tool} ok`,
      duration_ms: Date.now() - started,
    };
  }

  // 2. Fall through to LLM
  logger.debug(`[${rid}] no direct match — calling LLM`, rid);
  const resp  = await plan(text, rid);
  const calls = extractToolCalls(resp);
  if (!calls.length) {
    const text2 = (resp.message?.content || "").trim();
    logger.info(`[${rid}] no tool calls — assistant: ${text2}`, rid);
    return { request_id: rid, transcript: text, actions: [], assistant: text2, duration_ms: Date.now() - started };
  }

  const results = await Promise.all(calls.map(c => execOne(c.name, c.arguments, rid)));
  const errors  = results.filter(r => r.error);
  return {
    request_id: rid,
    transcript: text,
    actions: results,
    assistant: errors.length === 0
      ? (results.map(r => r.tool).join(", ") || "ok")
      : `${results.length - errors.length}/${results.length} succeeded`,
    duration_ms: Date.now() - started,
  };
}

function toolStatus() {
  return Object.entries(COMMAND_MAP).map(([toolName, translate]) => {
    let command = toolName;
    try {
      command = translate({ tempo: 120, enabled: true, index: -1, track_index: 0, volume_db: -6, pan: 0, muted: true, soloed: true, armed: true, clip_index: 0, scene_index: 0, start: 0, end: 4, time: 0 })[0];
    } catch {}
    return { tool: toolName, ableton_command: command, upstream_supported: UPSTREAM_ABLETON_COMMANDS.has(command) };
  });
}

module.exports = { processAudio, processCommand, toolStatus, COMMAND_MAP, UPSTREAM_ABLETON_COMMANDS, tryDirectParse };
